Kitkat compatibility

drodda opened the pull request:
Modified SystemClock class to support function NEXT_ALARM_FORMATTED for API<21. Still uses getNextAlarmClock() for API>=21

drodda committed:
Kitkat compatibility
Modified SystemClock class to support function NEXT_ALARM_FORMATTED for API<21. Still uses getNextAlarmClock() for API>=21

eladnava commented on app/src/main/java/com/eladnava/sunriser/alarms/SystemClock.java:
```diff
@@ -2,26 +2,73 @@
-
-        // No scheduled alarm?
-        if ( nextAlarm == null )
+        if(android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.LOLLIPOP)
```
Please import android.os =)

eladnava commented on app/src/main/java/com/eladnava/sunriser/alarms/SystemClock.java:
```diff
@@ -2,26 +2,73 @@
+
+            // Acquire an instance of the system-wide alarm manager
+            AlarmManager alarmManager = SystemServices.getAlarmManager(context);
+            // Query for next alarm info
```
Please add a line break before each comment =)

eladnava commented on app/src/main/java/com/eladnava/sunriser/alarms/SystemClock.java:
```diff
@@ -2,26 +2,73 @@
+            DateFormat dateFormat = new SimpleDateFormat(format, java.util.Locale.getDefault() );
+            try {
+                alarmDate = dateFormat.parse(nextAlarm);
+            }catch (java.text.ParseException e) {
```
Please import java.text;

eladnava commented:
@drodda Excellent work! I had a few comments, please address them and then we should be good to merge! 👍

eladnava commented on app/src/main/java/com/eladnava/sunriser/alarms/SystemClock.java:
```diff
@@ -2,26 +2,73 @@
+            // Get day, hour and minute for next alarm as a Date object
+            Date alarmDate;
+            String format = android.text.format.DateFormat.is24HourFormat(context) ? "E k:mm" : "E h:mm aa";
+            DateFormat dateFormat = new SimpleDateFormat(format, java.util.Locale.getDefault() );
```
Please import java.util;

drodda committed:
Code cleanup: added imports to SystemClock.java

drodda committed:
KitKat Compatibility:
API<21 does not support ACTION_NEXT_ALARM_CLOCK_CHANGED event.
Added class CheckSystemAlarm.java that periodically calls SunriseScheduler.rescheduleSunriseAlarm.

drodda committed:
* KitKat Compatibility:
Modified SunriseScheduler - stores previously used nextAlarm value, and only reschedules the alarm if new nextAlarm has changed.

drodda committed:
KitKat Compatibility:
Modified SunriseAlarm - now checks if an alarm should be performed before starting AsyncSunriseAlarm

drodda committed:
Bug fix: CheckSystemAlarm timer was not firing. Added permissions to AndroidManifest and changed CheckSystemAlarm to BroadcastReceiver.

drodda committed:
Moved CheckSystemAlarmScheduler functions to CheckSystemAlarm.
Added SimpleNotify class to display display debugging information (will be removed later)

drodda committed:
* Settings.java: Now disables CheckSystemAlarm task when sunrise is disabled, and starts it when sunrise is enabled.
* Removed WiFi requirement from startup (Main.java) so it can be run without WiFi
* Cleanup in SunriseScheduler.java - nextAlarmOld is being lost between alarms, so removed. No need to check if the alarm has changed, just reschedule it anyway.

drodda committed:
SunriseAlare: Stops CheckSystemAlarm process before beginning a sunrise, and resumes it after to prevent the sunrise from being interrupted.

drodda committed:
* Modified ThreadUtils.sleepExact to use Thread.Sheep instead of loop
* SunriseAlarm implements WakeLock to ensure timing while a sunrise is happening
* MiLightIntegration.fadeOutLightByZone repeated 5 times to ensure light turns off. May not be necessary.

eladnava commented:
Closing as this PR is outdated. Let me know if you'd like to try again. 👍
